hazelcast/src/main/java/com/hazelcast/internal/metrics/impl/GaugeImpl.java

```diff
@@ -0,0 +1,150 @@
+    volatile S source;
+
+    private final ILogger logger;
+    private final String name;
+
+    public GaugeImpl(String name, ILogger logger) {
+        this.name = name;
+        this.logger = logger;
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public void render(StringBuilder sb) {
+        checkNotNull(sb, "sb can't be null");
+
+        sb.append(name).append('=');
+
+        Object input = this.input;
+        S source = this.source;
+
+        if (input == null || source == null) {
+            sb.append("NA");
+            return;
+        }
+
+        try {
+            if (input instanceof LongProbe) {
+                LongProbe function = (LongProbe) input;
+                sb.append(function.get(source));
+            } else if (input instanceof DoubleProbe) {
+                DoubleProbe function = (DoubleProbe) input;
+                sb.append(function.get(source));
+            } else if (input instanceof AccessibleObjectGauge) {
+                AccessibleObjectGauge accessibleObjectInput = (AccessibleObjectGauge) input;
+                if (accessibleObjectInput.isDouble()) {
+                    sb.append(accessibleObjectInput.getDouble(source));
+                } else {
+                    sb.append(accessibleObjectInput.getLong(source));
+                }
+            }
+        } catch (Exception e) {
+            logger.warning("Failed to update metric:" + name, e);
+            sb.append("NA");
+        }
+    }
+
+    @Override
+    public long readLong() {
+        Object input = this.input;
+        S source = this.source;
+        long result = 0;
+
+        if (input == null || source == null) {
+            return result;
+        }
+
+        try {
+            if (input instanceof AccessibleObjectGauge) {
+                AccessibleObjectGauge accessibleObjectInput = (AccessibleObjectGauge) input;
+                if (accessibleObjectInput.isDouble()) {
+                    double doubleResult = accessibleObjectInput.getDouble(source);
+                    result = Math.round(doubleResult);
+                } else {
+                    result = accessibleObjectInput.getLong(source);
+                }
+            } else if (input instanceof LongProbe) {
+                LongProbe<S> longInput = (LongProbe<S>) input;
+                result = longInput.get(source);
+            } else if (input instanceof DoubleProbe) {
+                DoubleProbe<S> doubleInput = (DoubleProbe<S>) input;
+                double doubleResult = doubleInput.get(source);
+                result = Math.round(doubleResult);
+            }
+            return result;
+        } catch (Exception e) {
+            logger.warning("Failed to update metric:" + name, e);
+            return result;
+        }
+    }
+
+    @Override
+    public double readDouble() {
+        Object input = this.input;
+        S source = this.source;
+        double result = 0;
```

Comment:
This is a cleanup with the FieldProbe either dealing with Long or dealing with Double.

```
abstract class FieldProbe<S> {

    static <S> FieldProbe<S> create(Field field, Probe  probe) {
        int type = AccessibleObjectProbe.getType(field.getType());
        if (type == -1) {
            throw new IllegalArgumentException(format("@Gauge field '%s' is of an unhandled type", field));
        }

        if (AccessibleObjectProbe.isDouble(type)) {
            return new DoubleFieldProbe<S>(field, probe, type);
        } else {
            return new LongFieldProbe<S>(field, probe, type);
        }
    }

    final Probe probe;
    final Field field;
    final int type;

    FieldProbe(Field field, Probe probe, int type) {
        this.field = field;
        this.probe = probe;
        this.type = type;
        field.setAccessible(true);
    }

     void register(MetricsRegistryImpl metricsRegistry, Object source, String namePrefix) {
        String name = getName(namePrefix);
        metricsRegistry.registerInternal(source, name, this);
    }

    private String getName(String namePrefix) {
        String name = field.getName();
        if (!probe.name().equals("")) {
            name = probe.name();
        }

        return namePrefix + "." + name;
    }

    static class LongFieldProbe<S> extends FieldProbe implements LongProbe<S> {

        public LongFieldProbe(Field field, Probe probe, int type) {
            super(field, probe, type);
        }

        @Override
        public long get(S source) throws Exception {
            switch (type) {
                case TYPE_PRIMITIVE_LONG:
                    return field.getLong(source);
                case TYPE_LONG_NUMBER:
                    Number longNumber = (Number) field.get(source);
                    return longNumber == null ? 0 : longNumber.longValue();
                case TYPE_MAP:
                    Map<?, ?> map = (Map<?, ?>) field.get(source);
                    return map == null ? 0 : map.size();
                case TYPE_COLLECTION:
                    Collection<?> collection = (Collection<?>) field.get(source);
                    return collection == null ? 0 : collection.size();
                case TYPE_COUNTER:
                    Counter counter = (Counter) field.get(source);
                    return counter == null ? 0 : counter.get();
                default:
                    throw new IllegalStateException("Unhandled type:" + type);
            }
        }
    }

   static class DoubleFieldProbe<S> extends FieldProbe implements DoubleProbe<S> {

        public DoubleFieldProbe(Field field, Probe probe, int type) {
            super(field, probe, type);
        }

        @Override
        public double get(Object source) throws Exception {
            switch (type) {
                case TYPE_DOUBLE_PRIMITIVE:
                    return field.getDouble(source);
                case TYPE_DOUBLE_NUMBER:
                    Number doubleNumber = (Number) field.get(source);
                    return doubleNumber == null ? 0 : doubleNumber.doubleValue();
                default:
                    throw new IllegalStateException("Unhandled type:" + type);
            }
        }
    }
}
```

This allows the following cleanup:

```
  @Override
    public long readLong() {
        Object input = this.input;
        S source = this.source;
        long result = 0;

        if (input == null || source == null) {
            return result;
        }

        try {
           if (input instanceof LongProbe) {
                LongProbe<S> longInput = (LongProbe<S>) input;
                result = longInput.get(source);
            } else if (input instanceof DoubleProbe) {
                DoubleProbe<S> doubleInput = (DoubleProbe<S>) input;
                double doubleResult = doubleInput.get(source);
                result = Math.round(doubleResult);
            }
            return result;
        } catch (Exception e) {
            logger.warning("Failed to update metric:" + name, e);
            return result;
        }
    }
```

Since we don't need to deal with AccessibleObjectProbe anymore